dht-cache/src/cache/local.rs

```diff
@@ -0,0 +1,320 @@
+//! Local in-memory cache
+
+pub use crate::data::*;
+use crate::domopersistentstorage::{DomoPersistentStorage, SqlxStorage};
+use serde_json::Value;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::BTreeMap;
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+use tokio::sync::{OwnedRwLockReadGuard, RwLock};
+
+#[derive(Default)]
+pub(crate) struct InnerCache {
+    pub mem: BTreeMap<String, BTreeMap<String, DomoCacheElement>>,
+    pub store: Option<SqlxStorage>,
+}
+
+/// SAFETY: the SqlxStorage access is only over write()
+unsafe impl std::marker::Sync for InnerCache {}
+
+impl InnerCache {
+    pub fn put(&mut self, elem: &DomoCacheElement) {
+        let topic_name = elem.topic_name.clone();
+        let topic_uuid = &elem.topic_uuid;
+
+        self.mem
+            .entry(topic_name)
+            .and_modify(|topic| {
+                topic.insert(topic_uuid.to_owned(), elem.to_owned());
+            })
+            .or_insert_with(|| [(topic_uuid.to_owned(), elem.to_owned())].into());
+    }
+}
+
+/// Local cache
+#[derive(Default, Clone)]
+pub struct LocalCache(Arc<RwLock<InnerCache>>);
+
+impl LocalCache {
+    /// Instantiate a local cache from the configuration provided
+    ///
+    /// If url is empty do not try to bootstrap the cache from the db
+    ///
+    /// TODO: propagate errors
+    pub async fn with_config(db_config: &sifis_config::Cache) -> Self {
+        let mut inner = InnerCache::default();
+
+        if !db_config.url.is_empty() {
+            let mut store = SqlxStorage::new(db_config).await;
+
+            for a in store.get_all_elements().await {
+                inner.put(&a);
+            }
+
+            if db_config.persistent {
+                inner.store = Some(store);
+            }
+        }
+
+        LocalCache(Arc::new(RwLock::new(inner)))
+    }
+
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Feeds a slice of this type into the given [`Hasher`].
+    pub async fn hash<H: Hasher>(&self, state: &mut H) {
+        let cache = &self.0.read().await.mem;
+        for (topic_name, map_topic_name) in cache.iter() {
+            topic_name.hash(state);
+
+            for (topic_uuid, value) in map_topic_name.iter() {
+                topic_uuid.hash(state);
+                value.to_string().hash(state);
```

Comment:
Is there a way to easily split these comments into a separated issue? Obviously, for me it is easy to _just check the diff_, but I agree that the aim here is a bit different. If we can easily split issues that should be fixed in different PRs, it would be great.

Reply:
Sadly there isn't, it would be a nice feature though.

I'd do a round of performance improvements after we are good with the API rework :)